Description: add v4l2 access_demux AYUV fmt support.
 .
 vlc (3.0.18-2) unstable; urgency=medium
 .
Author: Som Qin <som.qin@starfivetech.com>

---
The information above should follow the Patch Tagging Guidelines, please
checkout https://dep.debian.net/deps/dep3/ to learn about the format. Here
are templates for supplementary fields that you might want to add:

Origin: (upstream|backport|vendor|other), (<patch-url>|commit:<commit-id>)
Bug: <upstream-bugtracker-url>
Bug-Debian: https://bugs.debian.org/<bugnumber>
Bug-Ubuntu: https://launchpad.net/bugs/<bugnumber>
Forwarded: (no|not-needed|<patch-forwarded-url>)
Applied-Upstream: <version>, (<commit-url>|commit:<commid-id>)
Reviewed-By: <name and email of someone who approved/reviewed the patch>
Last-Update: 2024-12-18

--- vlc-3.0.18.orig/modules/access/v4l2/demux.c
+++ vlc-3.0.18/modules/access/v4l2/demux.c
@@ -149,6 +149,7 @@ static const vlc_v4l2_fmt_t v4l2_fmts[]
 //  { V4L2_PIX_FMT_NV61,    VLC_CODEC_NV61, 1, 0, 0, 0 },
     { V4L2_PIX_FMT_NV12,    VLC_CODEC_NV12, 1, 0, 0, 0 },
     { V4L2_PIX_FMT_NV21,    VLC_CODEC_NV21, 1, 0, 0, 0 },
+    { V4L2_PIX_FMT_AYUV32,  VLC_CODEC_VUYA, 4, 0, 0, 0 },
 
     /* V4L2-documented but VLC-unsupported misc. YUV formats */
 //  { V4L2_PIX_FMT_Y41P       },
@@ -222,6 +223,8 @@ static const vlc_v4l2_fmt_t v4l2_fmts[]
     { V4L2_PIX_FMT_GREY,    VLC_CODEC_GREY, 1, 0, 0, 0 },
 };
 
+int frame_get_half = 0;
+
 static const vlc_v4l2_fmt_t *vlc_from_v4l2_fourcc (uint32_t fourcc)
 {
      for (size_t i = 0; i < sizeof (v4l2_fmts) / sizeof (v4l2_fmts[0]); i++)
@@ -551,6 +554,10 @@ static int InitVideo (demux_t *demux, in
              es_fmt.video.i_sar_den);
     sys->es = es_out_Add (demux->out, &es_fmt);
 
+    frame_get_half =  ((es_fmt.i_codec == VLC_CODEC_VUYA)
+             && (es_fmt.video.i_frame_rate / es_fmt.video.i_frame_rate_base > 29)
+             && (es_fmt.video.i_width * es_fmt.video.i_height > 1280 *720)) ? 1 : 0;
+
     /* Init I/O method */
     void *(*entry) (void *);
     if (caps & V4L2_CAP_STREAMING)
--- vlc-3.0.18.orig/modules/access/v4l2/linux/videodev2.h
+++ vlc-3.0.18/modules/access/v4l2/linux/videodev2.h
@@ -546,6 +546,8 @@ struct v4l2_pix_format {
 #define V4L2_PIX_FMT_YUV420M v4l2_fourcc('Y', 'M', '1', '2') /* 12  YUV420 planar */
 #define V4L2_PIX_FMT_YVU420M v4l2_fourcc('Y', 'M', '2', '1') /* 12  YVU420 planar */
 
+#define V4L2_PIX_FMT_AYUV32  v4l2_fourcc('A', 'Y', 'U', 'V') /* 32  AYUV-8-8-8-8  */
+
 /* Bayer formats - see http://www.siliconimaging.com/RGB%20Bayer.htm */
 #define V4L2_PIX_FMT_SBGGR8  v4l2_fourcc('B', 'A', '8', '1') /*  8  BGBG.. GRGR.. */
 #define V4L2_PIX_FMT_SGBRG8  v4l2_fourcc('G', 'B', 'R', 'G') /*  8  GBGB.. RGRG.. */
--- vlc-3.0.18.orig/modules/access/v4l2/video.c
+++ vlc-3.0.18/modules/access/v4l2/video.c
@@ -582,6 +582,8 @@ mtime_t GetBufferPTS (const struct v4l2_
 /*****************************************************************************
  * GrabVideo: Grab a video frame
  *****************************************************************************/
+extern int frame_get_half;
+static int half = 0;
 block_t *GrabVideo (vlc_object_t *demux, int fd,
                     const struct buffer_t *restrict bufv)
 {
@@ -605,6 +607,10 @@ block_t *GrabVideo (vlc_object_t *demux,
                 return NULL;
         }
     }
+    if (frame_get_half && half++ % 2) {
+        v4l2_ioctl (fd, VIDIOC_QBUF, &buf);
+        return NULL;
+    }
 
     /* Copy frame */
     block_t *block = block_Alloc (buf.bytesused);
